doc/design/switch_kernel.md

```diff
@@ -0,0 +1,66 @@
+## Background
+Every operator has many kernels. It because there are multiple data types, places, data layout that Fluid supports. We use the `KernelType` to describe kernel types that operators can hold. 
+
+The `KernelType` is as follows.
+
+```
+struct KernelType {
+  Place place_;
+  DataType data_type_;
+  LayoutType layout_;
+};
+```
+
+The `place_` is a descriptor of the device and the computational library, e.g., `MKLPlace`, `CUDAPlace`.
+
+The `data_type_` is the data type that this kernel performs on, e.g., `FP32`, `INT64`. Note that one kernel may have inputs with different data types. However, it will be a major `data_type`. For example, the `cross_entropy` takes `int64` as it label, and `double`/`float` as its input logit and output cost. The major `data_type` of `cross_entropy` is `float`/`double`.
```

Comment:
do we need a uniform way to set which input is used to determine the major `data_type`, this can be done in `AddInput` of OpProtoAndCheckerMaker 
like `AsKernelTypeDecider`

Reply:
It is another topic.

However, operator developers may be familiar with plain C++, not a register syntax.